// dice-js/die.js

var Proteus = require("proteus"),
    Die
;

module.exports = Die = Proteus.Class.derive(Object.defineProperties({
    
    self: {
        initialize: function (instance, args) {
            var sides = args[0],
                dice = this.__dice__,
                die = dice[sides]
            ;
            
            if (die) {
                return die;
            }
            else {
                if (!sides || sides < 2 || typeof sides !== "number") {
                    throw "Argument to Die constructor must be a number greater than 1.";
                }
                dice[sides] = instance;
            }
        }
    },
    
    init: function (sides) {
        Object.defineProperties(this, {
            sides: { value: sides, enumerable: true }
        });
    },
    
    roll: function () {
        return Math.floor(Math.random() * this.sides) + 1;
    }
    
}, {
    valueOf: { value: function () {
        return this.roll();
    } },
    
    toString: { value: function () {
        return '[d' + this.sides + ']';
    } }
    
}));

Object.defineProperty(Die, "__dice__", { value: [] });
