// dice-js/expression.js

var Proteus = require("proteus");

module.exports = Proteus.Class.derive(Object.defineProperties({
    
    init: function () {
        Object.defineProperties(this, {
            members: { value: [], enumerable: true }
        });
        
        Proteus.slice(arguments).forEach(this.add, this);
    },
    
    add: function (exp) {
        return this.members.push(exp);
    },
    
    remove: function (idx) {
        return this.members.splice(idx, 1);
    },
    
    reset: function () {
        this.members.length = 0;
        return this;
    }
    
}, {
    length: {
        get: function () {
            return this.members.length;
        },
        set: function (v) {
            this.members.length = v;
        }
    },
    
    valueOf: { value: function () {
        var results = [],
            members = this.members,
            len     = members.length,
            i
        ;
        
        for (i = 0; i < len; i++) {
            results.push(members[i].valueOf());
        }
        
        return results.length <= 1 ? results[0] : results;
    } }
}));
